// @sanity/desk-tool/lib/components/DocumentStatusBar/DocumentStatusBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DocumentStatusBar;

var _react = _interopRequireDefault(require("react"));

var _DocumentStatusBar = _interopRequireDefault(require("./DocumentStatusBar.css"));

var _DocumentStatusBarActions = require("./DocumentStatusBarActions");

var _DocumentStatusBarBadges = require("./DocumentStatusBarBadges");

var _SyncState = require("./SyncState");

var _TimeAgo = _interopRequireDefault(require("../../components/TimeAgo"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DocumentStatusBar(props) {
  return _react.default.createElement("div", {
    className: _DocumentStatusBar.default.root
  }, _react.default.createElement("div", {
    className: _DocumentStatusBar.default.status
  }, _react.default.createElement(_DocumentStatusBarBadges.DocumentStatusBarBadges, {
    id: props.id,
    type: props.type
  }), _react.default.createElement("div", {
    className: _DocumentStatusBar.default.statusDetails
  }, _react.default.createElement("button", {
    className: _DocumentStatusBar.default.lastUpdatedButton,
    onClick: props.onLastUpdatedButtonClick,
    type: "button"
  }, props.lastUpdated ? _react.default.createElement(_react.default.Fragment, null, "Updated ", _react.default.createElement(_TimeAgo.default, {
    time: props.lastUpdated
  })) : 'Empty'), ' ', _react.default.createElement(_SyncState.SyncState, {
    id: props.id,
    type: props.type
  }))), _react.default.createElement("div", {
    className: _DocumentStatusBar.default.actions
  }, _react.default.createElement("div", {
    className: _DocumentStatusBar.default.actionsWrapper
  }, _react.default.createElement(_DocumentStatusBarActions.DocumentStatusBarActions, {
    id: props.id,
    type: props.type
  }))));
}
